winit-wayland/src/window/state.rs

```diff
@@ -1069,6 +1085,44 @@ impl WindowState {
         self.title = title;
     }
 
+    /// Set the window's icon
+    pub fn set_window_icon(&mut self, window_icon: Option<winit_core::icon::Icon>) {
+        let xdg_toplevel_icon_manager = match self.xdg_toplevel_icon_manager.as_ref() {
+            Some(xdg_toplevel_icon_manager) => xdg_toplevel_icon_manager,
+            None => {
+                warn!("`xdg_toplevel_icon_manager_v1` is not supported");
+                return;
+            },
+        };
+
+        let xdg_toplevel_icon = match window_icon {
+            Some(icon) => {
+                let mut image_pool = self.image_pool.lock().unwrap();
+                let toplevel_icon = match ToplevelIcon::new(icon, &mut image_pool) {
+                    Ok(toplevel_icon) => toplevel_icon,
+                    Err(error) => {
+                        warn!("Error setting window icon: {error}");
+                        return;
+                    },
+                };
+
+                let xdg_toplevel_icon =
+                    xdg_toplevel_icon_manager.create_icon(&self.queue_handle, GlobalData);
+
+                toplevel_icon.add_buffer(&xdg_toplevel_icon);
+                self.toplevel_icon = Some(toplevel_icon);
+
+                Some(xdg_toplevel_icon)
+            },
+            None => {
+                self.toplevel_icon = None;
+                None
+            },
+        };
+
+        xdg_toplevel_icon_manager.set_icon(self.window.xdg_toplevel(), xdg_toplevel_icon.as_ref());
```

Comment:
After looking a bit into protocol I've noticed that, it seems like this should be dropped after calling `set_icon`, since it behaves similar to `wl_region`. We also have to `destroy` the global, etc, etc.